javalib/src/main/scala/java/util/RedBlackTree.scala

```diff
@@ -0,0 +1,1033 @@
+          !isWithinLowerBound(node.key, end, endKind)) {
+        null
+      } else {
+        node
+      }
+    }
+  }
+
+  private[this] final class DescendingEntriesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, Map.Entry[A, B]](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): Map.Entry[A, B] = node
+  }
+
+  private[this] final class DescendingKeysIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, A](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): A = node.key
+  }
+
+  private[this] final class DescendingValuesIterator[A, B](tree: Tree[A, B],
+      start: A, startKind: BoundKind, end: A, endKind: BoundKind)(
+      implicit comp: Comparator[_ >: A])
+      extends DescendingTreeIterator[A, B, B](tree, start, startKind, end, endKind) {
+
+    def nextResult(node: Node[A, B]): B = node.value
+  }
+
+  // building
+
+  /** Build a Tree suitable for a TreeSet from an ordered sequence of keys. */
+  def fromOrderedKeys[A](xs: Iterator[A], size: Int): Tree[A, Any] = {
+    val maxUsedDepth = 32 - Integer.numberOfLeadingZeros(size) // maximum depth of non-leaf nodes
+
+    def f(level: Int, size: Int): Node[A, Any] = size match {
+      case 0 =>
+        null
+      case 1 =>
+        new Node(xs.next(), (), level == maxUsedDepth && level != 1, null, null, null)
+      case n =>
+        val leftSize = (size - 1) >> 1 // (size - 1) / 2
+        val left = f(level + 1, leftSize)
+        val x = xs.next()
+        val right = f(level + 1, size - 1 - leftSize)
+        val n = new Node(x, (), false, left, right, null)
+        if (left ne null)
+          left.parent = n
+        right.parent = n
+        n
+    }
+
+    new Tree(f(1, size), size)
+  }
+
+  /** Build a Tree suitable for a TreeMap from an ordered sequence of key/value pairs. */
+  def fromOrderedEntries[A, B](xs: Iterator[Map.Entry[A, B]], size: Int): Tree[A, B] = {
+    val maxUsedDepth = 32 - Integer.numberOfLeadingZeros(size) // maximum depth of non-leaf nodes
+
+    def f(level: Int, size: Int): Node[A, B] = size match {
+      case 0 =>
+        null
+      case 1 =>
+        val entry = xs.next()
+        new Node(entry.getKey(), entry.getValue(),
+            level == maxUsedDepth && level != 1, null, null, null)
+      case n =>
+        val leftSize = (size - 1) >> 1 // (size - 1) / 2
+        val left = f(level + 1, leftSize)
+        val entry = xs.next()
+        val right = f(level + 1, size - 1 - leftSize)
+        val n = new Node(entry.getKey(), entry.getValue(), false, left, right, null)
+        if (left ne null)
+          left.parent = n
+        right.parent = n
+        n
+    }
+
+    new Tree(f(1, size), size)
+  }
```

Comment:
You should share these implementations. Use an accessor function + @inline and it will be zero overhead.